### PR TITLE
DRILL-4203: Parquet File. Date is stored wrongly

### PR DESCRIPTION
**Pull request by vdiravka**
Drill was writing non-standard dates into parquet files for all releases
before this commit. The values have been read correctly by Drill, but
external tools like Spark reading the files will see corrupted values for
all dates that have been written by Drill.

This change corrects the behavior of the Drill parquet writer to correctly
store dates in the format given in the parquet specification.

To maintain compatibility with old files, the parquet reader code has been
updated to check for the old format and automatically shift the
corrupted values into corrected ones automatically.

The test cases included here should ensure that all files produced by
historical versions of Drill will continue to return the same values
they had in previous releases. For compatibility with external tools, any
old files with corrupted dates can be re-written using the CREATE TABLE AS
command (as the writer will now only produce the specification-compliant
values, even if after reading out of older corrupt files, one
new extra field "is.date.correct = true" will be included into the parquet meta
information of files and into drill metadata cache files).

While the old behavior was a consistent shift into an unlikely range to be
used in a modern database (over 10,000 years in the future), these are
still valid date values. In the case where these may have been written
into files intentionally, and we cannot be certain from the metadata if
Drill produced the files, an option is included to turn off the auto-correction.
Use of this option is assumed to be extremely unlikely, but it is included for
completeness.

One small fix in the ParquetGroupScan to accommodate changes in master that changed
when metadata is read.

Added new tests for bugs (revealed by the regression suite) with old and new
parquet (binary) files for new tests, updated metadata cache files accordingly.

Removed unnecessary double conversion of value with Julian day.

Added ability to correct corrupted dates for parquet files with the second
version old metadata cache file as well.

Fix DrillVersionInfo to make it provide a valid version number even during
the unit tests. This is now a build-time generated class, rather than one
that looks on the classpath for META-INF files. (This pattern
for file generation with parameters passed from the POM files
was borrowed from parquet-mr)

### CODE REVIEW

**Comment by vdiravka** [comment]
@jaltekruse Could you please review?

**Comment by parthchandra** [comment]
@vdiravka can you repost this PR with the commits split into Jason's original work and then with your fixes on top? We should give credit where it is due.

**Comment by vdiravka** [comment]
@parthchandra Done. Now it involves 2 Jason original commits (only with resolved git version conflicts) and one my commit on top.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/columnreaders/ParquetRecordReader.java [hunk 1]
```diff
@@ -207,6 +229,31 @@ public OperatorContext getOperatorContext() {
     return operatorContext;
   }
 
+  /**
```
Comment (parthchandra) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/columnreaders/ParquetRecordReader.java]: This seems completely unrelated to the bug

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/Metadata.java [hunk 2]
```diff
@@ -918,18 +916,22 @@ public void setMax(Object max) {
     @JsonProperty public ConcurrentHashMap<ColumnTypeMetadata_v2.Key, ColumnTypeMetadata_v2> columnTypeInfo;
     @JsonProperty List<ParquetFileMetadata_v2> files;
     @JsonProperty List<String> directories;
-    @JsonProperty String drillVersion;
```
Comment (jaltekruse) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/Metadata.java]: I had intentionally added the drill version here assuming that it would be good information to have around if a similar issue ever comes up the the future, as well as provide all of the info we need to have an explicit flag that the dates have become correct. For this to work completely, this commit should be the last commit right before a release (it could be a point release). Any particular reason that we would want to not write it into the file?

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/Metadata.java [hunk 3]
```diff
@@ -935,6 +972,11 @@ public ColumnTypeMetadata_v2 getColumnTypeInfo(String[] name) {
     @JsonIgnore @Override public ParquetTableMetadataBase clone() {
       return new ParquetTableMetadata_v2(files, directories, columnTypeInfo);
     }
+
+    @JsonIgnore @Override public boolean isDateCorrect() {
+      return isDateCorrect;
```
Comment (vdiravka) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/Metadata.java]: If metadata cache file is existed Drill reads it instead of retrieving metadata from multiple Parquet files. In the case when it was generated with drill after this commit the value of isDateCorrect will be true. In the case when it was generated with drill before this commit the isDateCorrect field in metadata cache file will be absent and value of this will be false in ParquetTableMetadata_v2.
And according to this value we just define DateCorruptionStatus (you can see more in ParquetReaderUtility.correctDatesInMetadataCache()). The leftover way of data checking in the cache was not changed.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetGroupScan.java [hunk 4]
```diff
@@ -664,27 +667,37 @@ private void init(MetadataContext metaContext) throws IOException {
...
       Path metaPath = new Path(p, Metadata.METADATA_FILENAME);
       if (fs.isDirectory(new Path(selectionRoot)) && fs.exists(metaPath)) {
         usedMetadataCache = true;
         if (parquetTableMetadata == null) {
-          parquetTableMetadata = Metadata.readBlockMeta(fs, metaPath.toString(), metaContext);
+          parquetTableMetadata = Metadata.readBlockMeta(fs, metaPath.toString(), metaContext, formatConfig);
         }
         if (fileSet != null) {
-          parquetTableMetadata = removeUnneededRowGroups(parquetTableMetadata);
+          if (parquetTableMetadata == null) {
```
Comment (vdiravka) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetGroupScan.java]: agree. It is redundant. Changed in the last commit.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/columnreaders/ParquetRecordReader.java (hunk 1 above)
Comment (vdiravka) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/columnreaders/ParquetRecordReader.java]: It is just Jason's minor refactoring. Is it ok?

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/Metadata.java (hunk 2 above)
Comment (vdiravka) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/Metadata.java]: I agree regarding future similar issues. So I returned this field to `ParquetTableMetadataBase` in the new commit. 
But to determine that the parquet file is new and definitely with correct date values we have a new flag in parquet metadata "is.date.correct = true". Thus, it is more easy to determine corrupted date values and there is no need to wait the end of release to merge these changes.

File: contrib/storage-hive/core/src/main/java/org/apache/drill/exec/store/hive/HiveDrillNativeScanBatchCreator.java [hunk 5]
```diff
@@ -123,6 +123,7 @@ public ScanBatch getBatch(FragmentContext context, HiveDrillNativeParquetSubScan
           // in the first row group
           ParquetReaderUtility.DateCorruptionStatus containsCorruptDates =
               ParquetReaderUtility.detectCorruptDates(parquetMetadata, config.getColumns(), true);
+          logger.info(containsCorruptDates.toString());
```
Comment (jaltekruse) [inline contrib/storage-hive/core/src/main/java/org/apache/drill/exec/store/hive/HiveDrillNativeScanBatchCreator.java]: not a very useful logging statement to check in, should at least contain a message about what this value is. This will just print a boolean value. If it was just for local debugging it can be removed.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/Metadata.java (hunk 2 above)
Comment (jaltekruse) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/Metadata.java]: This sounds reasonable, having both is okay.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetReaderUtility.java [hunk 6]
```diff
@@ -184,17 +201,15 @@ public static DateCorruptionStatus detectCorruptDates(ParquetMetadata footer,
           if (parsedCreatedByVersion.hasSemanticVersion()) {
             SemanticVersion semVer = parsedCreatedByVersion.getSemanticVersion();
             String pre = semVer.pre + "";
-            if (semVer != null && semVer.major == 1 && semVer.minor == 8 && semVer.patch == 1 && pre.contains("drill")) {
```
Comment (jaltekruse) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetReaderUtility.java]: I don't know if you are guarding against a null value for semVer at a higher level, but I'm pretty sure the reason I added it is that older versions of parquet files can be lacking this field. I assume with all of the tests cases that were added we should be okay, but it might be better to keep it in defensively. If this field isn't set in the metadata, it shouldn't make it an invalid file, but if we get an NPE it will stop query processing.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetScanBatchCreator.java [hunk 7]
```diff
@@ -107,6 +107,7 @@ public ScanBatch getBatch(FragmentContext context, ParquetRowGroupScan rowGroupS
         boolean autoCorrectCorruptDates = rowGroupScan.formatConfig.autoCorrectCorruptDates;
         ParquetReaderUtility.DateCorruptionStatus containsCorruptDates = ParquetReaderUtility.detectCorruptDates(footers.get(e.getPath()), rowGroupScan.getColumns(),
                 autoCorrectCorruptDates);
+        logger.info(containsCorruptDates.toString());
```
Comment (jaltekruse) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetScanBatchCreator.java]: same as above, not a very useful logging statement.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetReaderUtility.java (hunk 6 above)
Comment (vdiravka) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetReaderUtility.java]: If `semVer `will be null we get NPE above, where `semVer.pre` is called. 
It happened with tests on our regression test framework. That's why I added `hasSemanticVersion()` above. So if this method return `true`, `semVer `cann't be `NULL` according `org.apache.parquet.SemanticVersion` and the following `semVer!=null` checking is redundant.

File: contrib/storage-hive/core/src/main/java/org/apache/drill/exec/store/hive/HiveDrillNativeScanBatchCreator.java (hunk 5 above)
Comment (vdiravka) [inline contrib/storage-hive/core/src/main/java/org/apache/drill/exec/store/hive/HiveDrillNativeScanBatchCreator.java]: But it is not a regular Boolean value, `DateCorruptionStatus` is an enum with overrided toString method. For example how log file looks after querying hive parquet file with correct date values:

`2016-10-06 23:32:30,598 [280920f1-e362-136e-0fdd-24779fef2c4a:frag:0:0] INFO  o.a.d.e.s.p.ParquetScanBatchCreator - It is determined from metadata that the date values are definitely CORRECT`

Is it enought?

Comment (jaltekruse) [inline contrib/storage-hive/core/src/main/java/org/apache/drill/exec/store/hive/HiveDrillNativeScanBatchCreator.java]: Sounds good

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetReaderUtility.java (hunk 6 above)
Comment (jaltekruse) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetReaderUtility.java]: Sorry I should have looked through the rest of the changes, thank you for fixing this

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetScanBatchCreator.java (hunk 7 above)
Comment (jaltekruse) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetScanBatchCreator.java]: addressed, in other similar comment, this is good

**Comment by vdiravka** [comment]
Thanks @jaltekruse.
@parthchandra Could you please take a look?

**Comment by vdiravka** [comment]
The branch is rebased to the master version.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/columnreaders/FixedByteAlignedReader.java [hunk 8]
```diff
@@ -161,7 +160,7 @@ void addNext(int start, int index) {
         intValue = readIntLittleEndian(bytebuf, start);
       }
 
-      mutator.set(index, DateTimeUtils.fromJulianDay(intValue + ParquetReaderUtility.CORRECT_CORRUPT_DATE_SHIFT));
+      mutator.set(index, (intValue - ParquetReaderUtility.CORRECT_CORRUPT_DATE_SHIFT) * DateTimeConstants.MILLIS_PER_DAY);
```
Comment (parthchandra) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/columnreaders/FixedByteAlignedReader.java]: Reviewing this math can give anyone a headache :(. But it is correct.

**Comment by tushu1232** [comment]
How can I get to test this fix .Thanks

**Comment by vdiravka** [comment]
@tushu1232 Until this fix is merged into the master you can clone my fork repository (https://github.com/vdiravka/drill), switch to the DRILL-4203 branch and build the project.
